swim/disseminator_test.go

```diff
@@ -356,3 +360,143 @@ func contains(cs []Change, address string) bool {
+	defer b.Destroy()
+
+	// after bootstrap a's memberlist contains a and b,
+	// but b's membership only contains b.
+	bootstrapNodes(t, b, a)
+
+	// clear changes so that a full sync will be issued
+	a.node.disseminator.ClearChanges()
+	b.node.disseminator.ClearChanges()
+
+	// check that indeed b's memberlist doesn't contain a
+	addrA := a.node.Address()
+	_, has := b.node.memberlist.Member(addrA)
+	assert.False(t, has, "expected that node1 is not yet part of a's membership")
+
+	cont := make(chan struct{})
+
+	// only when a reverse full sync is issued the membership of b changes
+	// we listen for a membership change and continue the main thread of the test
+	b.node.RegisterListener(onMembershipChange(func() {
+		cont <- struct{}{}
+	}))
+
+	// this forces a to ping b, b will respond with a full sync and
+	// perform a reverse full sync subsequently
+	a.node.pingNextMember()
+
+	// wait until the membership has changed and the cont channel is sent a signal
+	select {
+	case <-time.After(time.Second):
+		assert.Fail(t, "Test timed out, memberships did not converge")
+	case <-cont:
+	}
+
+	// check that b has indeed added a to its membership
+	_, has = b.node.memberlist.Member(addrA)
+	assert.True(t, has, "expected that a is part of b's membership now")
+}
+
+// onMembershipChange can be registered on an EventListener and fires f only
+// when the membership of this node changes.
+func onMembershipChange(f func()) ListenerFunc {
+	return ListenerFunc(func(e events.Event) {
+		switch e.(type) {
+		case MemberlistChangesAppliedEvent:
+			f()
+		}
+	})
+}
+
+// TestThrottleBidirectionalFullSyncs tests if we throttle the maximum number
+// of reverse full syncs running at the same time. In this test, we start
+// maxReverseFullSyncJobs+1 reverse full syncs and we block these routines midway
+// by waiting for the block channel. The last reverse full sync will be omitted
+// because the max number of reverse full syncs is in process (albeit blocked).
+// We wait and listen for the OmitReverseFullSyncEvent. When it fires, we
+// continue the test and count if the number of started reverse full syncs is
+// indeed equal to the maxBidirectionalFullSync variable.
+func TestThrottleBidirectionalFullSyncs(t *testing.T) {
+	tnode := newChannelNode(t)
+	maxJobs := tnode.node.maxReverseFullSyncJobs
+
+	tnodes := genChannelNodes(t, maxJobs+1)
+
+	bootstrapNodes(t, append(tnodes, tnode)...)
+	waitForConvergence(t, time.Second, append(tnodes, tnode)...)
+	block := make(chan struct{})
+	quit := make(chan struct{})
+	total := int64(0)
+
+	// Block the reverse full sync procedure to test if we throttle correctly.
+	tnode.node.RegisterListener(onStartReverseFullSync(func() {
+		atomic.AddInt64(&total, 1)
+		<-block
+	}))
+
+	// Listen for a reverse full sync that is omitted because the max number of
+	// reverse full sync routines are running.
+	tnode.node.RegisterListener(onOmitReverseFullSync(func() {
+		go func() {
+			for i := 0; i < maxJobs; i++ {
```

Comment:
My concern is that if we have a off-by-one error, and we omit earlier than expected, `onOmitReverseFullSync` will block when trying to free an additional `onStartReverseFullSync`. Because `onOmitReverseFullSync` is not called in a goroutine (neither by the event system or by the test code), it can potentially run before the last `select` in the test, deadlocking the test. I think we should either use `total` or, even better, add a buffer to the `block` channel equal to `maxJobs`.